openmdao/drivers/differential_evolution_driver.py

```diff
@@ -0,0 +1,483 @@
+import copy
+import numpy as np
+import openmdao
+
+from openmdao.core.analysis_error import AnalysisError
+from openmdao.core.driver import Driver, RecordingDebugging
+from openmdao.utils.mpi import MPI
+from six import iteritems, itervalues, next
+
+try:
+    from tqdm import tqdm
+except ModuleNotFoundError:
+
+    def tqdm(i, _):
+        return i
+
+
+from differential_evolution import DifferentialEvolution, EvolutionStrategy
+
+if not MPI:
+    rank = 0
+else:
+    rank = MPI.COMM_WORLD.rank
+
+
+class DifferentialEvolutionDriver(Driver):
+    def __init__(self, **kwargs):
+        """
+        Initialize the DifferentialEvolution driver.
+
+        Parameters
+        ----------
+        **kwargs : dict of keyword arguments
+            Keyword arguments that will be mapped into the Driver options.
+        """
+        super(DifferentialEvolutionDriver).__init__(**kwargs)
+
+        # What we support
+        self.supports["integer_design_vars"] = True
+        self.supports["inequality_constraints"] = True
+        self.supports["equality_constraints"] = True
+        self.supports["multiple_objectives"] = True
+
+        # What we don't support yet
+        self.supports["two_sided_constraints"] = False
+        self.supports["linear_constraints"] = False
+        self.supports["simultaneous_derivatives"] = False
+        self.supports["active_set"] = False
+
+        self._desvar_idx = {}
+        self._es = None
+        self._de = None
+
+        # Support for Parallel models.
+        self._concurrent_pop_size = 0
+        self._concurrent_color = 0
+
+    def _declare_options(self):
+        """
+        Declare options before kwargs are processed in the init method.
+        """
+        self.options.declare(
+            "strategy",
+            default="rand-to-best/1/exp/random",
+            desc="Evolution strategy to use for the differential evolution.",
+        )
+        self.options.declare(
+            "Pm",
+            desc="Mutation rate.",
+            default=None,
+            lower=0.0,
+            upper=1.0,
+            allow_none=True,
+        )
+        self.options.declare(
+            "Pc",
+            default=None,
+            lower=0.0,
+            upper=1.0,
+            allow_none=True,
+            desc="Crossover rate.",
+        )
+        self.options.declare(
+            "adaptivity", default=2, values=[0, 1, 2], desc="Self-adaptivity setting."
+        )
+        self.options.declare(
+            "max_gen", default=1000, desc="Number of generations before termination."
+        )
+        self.options.declare(
+            "tolx", default=1e-8, desc="Tolerance of the design vectors' spread."
+        )
+        self.options.declare(
+            "tolf", default=1e-8, desc="Tolerance of the fitness spread."
+        )
+        self.options.declare(
+            "pop_size",
+            default=0,
+            desc="Number of points in the GA. Set to 0 and it will be computed "
```

Comment:
Oops. Yeah, like I said: I reused code from the SimpleGADriver class because it works in much the same way. As I mentioned, eventually it may be good to abstract some of the logic out to reduce code duplication. But for now I think it would be better to keep the two as is, if you agree. But I will certainly take another good pass on removing stuff from SimpleGA which does not apply to the new class.

Reply:
Addressed by 392211fcf4110e866b6dcd95c110aabb60695e72